### PR TITLE
Posting to readwise

### PR DESCRIPTION
**Pull request by MKSG-MugunthKumar**
It's getting late here in Australia.
The code works, but I haven't tested it thoroughly.
I also refactored and moved methods around.
Please let me know if this looks good.

### CODE REVIEW

**Comment by MKSG-MugunthKumar** [comment]
also, the speech recognition is not very good. May be, should we consider using `openai-whisper`?

File: notion.py [hunk 1]
```diff
@@ -2,43 +2,43 @@
 import os
 from datetime import datetime
 
+class NotionExporter:
```
Comment (gostak-dd) [inline notion.py:5]: any reason why this is now a class?

File: command.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            if kwarg == "":
+                continue
+            li_kwarg = kwarg.split("=")
+
+            if not len(li_kwarg) > 1:
+                await self.invalid_kwarg_callback()
+
+            _kwargs[li_kwarg[0]] = li_kwarg[1]
+
+    if not self.audible_obj and command not in AUTHLESS_COMMANDS:
```
Comment (gostak-dd) [inline command.py:69]: why did you change this if? Now running commands like list_books without activation bytes crash the script.

Please run through everything again without activation bytes, without auth etc.

Remember a lot of users are not technical people so we have to really cover edge cases like this and make it easy to use

**Comment by MKSG-MugunthKumar** [comment]
@GGyll Mate, do you have some time to look into this, please?

**Comment by gostak-dd** [comment]
@MugunthKumar Hey man I left some comments did you not see them?

Overall looks good but some changes made the software crash for me without authenticating.

Also I am open to openai-whisper but do you need an account for that and a paid plan?
The current speech recognition isn't that good but it is free, and the user doesn't have to create an account and use their API key which is a bonus

**Comment by MKSG-MugunthKumar** [comment]
I don't see any comments, mate. Can you link them here, please?
I will have a look at the crash over the weekend.

**Comment by gostak-dd** [comment]
image
if you open the PR you should see them under the commits

**Comment by MKSG-MugunthKumar** [comment]
I think you have to publish the comments so I can see it.

**Review by gostak-dd — CHANGES_REQUESTED** [review]

**Comment by gostak-dd** [comment]
Oops

**Comment by MKSG-MugunthKumar** [comment]
Got it now 😄—MugunthOn 3 Aug 2024, at 5:23 pm, Gustaf ***@***.***> wrote:﻿
Oops

File: notion.py (hunk 1 above)
Comment (MKSG-MugunthKumar) [reply notion.py:5]: I'm modelling all exporters based on Readwise. They hold the necessary data (or initialized from command.py) instead of fetching it from an environment variable.

(Although it is not implemented)

Only Readwise is done as of now.

File: command.py (hunk 2 above)
Comment (MKSG-MugunthKumar) [reply command.py:69]: Oops. The bug is now fixed. It was in the way the object is initialized in command.py.

**Comment by MKSG-MugunthKumar** [comment]
Sorry about the delay. I was held up in my work. Please give it a go and let me know.

**Comment by gostak-dd** [comment]
The commands dont work, i try to run authenticate and get invalid_command_callback().
When debugging, self.audible_obj is null

This is on a fresh clone where i dont have any auth or anything

**Comment by gostak-dd** [comment]
You are setting self.audible_obj conditionally if there is already an auth, it needs to be set either way, or the program fails. Refer to main_file.py in the original repo

**Comment by gostak-dd** [comment]
Hey man any updates on this? Overall, I like the cleanup and code splitting you've done but the bugs need to be sorted out..

**Comment by MKSG-MugunthKumar** [comment]
Ouch, sorry mate. I completely forgot this.I’ll get this over tomorrow.—MugunthOn 11 Sep 2024, at 1:21 am, Gustaf ***@***.***> wrote:﻿
Hey man any updates on this? Overall, I like the cleanup and code splitting you've done but the bugs need to be sorted out..

**Comment by MKSG-MugunthKumar** [comment]
Hey,
This fix is now done. Sorry about the delay.

**Comment by MKSG-MugunthKumar** [comment]
I see some merge conflicts, mostly around formatting.

When you merge my changes, do a `--theirs` change for `audible_api.py` and `--ours` for `Readme.md`
Re-run the formatting tool to reformat my version of `audible_api.py` that you got from the merge.

**Comment by gostak-dd** [comment]
Looking good, can you ensure the clips and transcriptions appear in a folder in the same directory as the cloned repo? For me it creates a directory called 'audioextractor' in my home directory

**Comment by gostak-dd** [comment]
artifacts_root_directory ^

**Comment by gostak-dd** [comment]
Where are the highlights supposed to go in Readwise? I run the command and send a successful request and confirmed I am sending highlights, but I can't see them anywhere. I just signed up for Readwise so not sure how it works.

**Comment by MKSG-MugunthKumar** [comment]
`artifacts_root_directory` points to `~/audioextractor`. Feel free to change that. I moved all artifacts generated by the code to this directory because it is writable by all users (non-root and non-admin) on all OSes.

Let's say if a user installs this app in `/usr/bin`, the app is going to crash unless the user runs it using sudo -- not a great experience.

The value of this is controlled by a variable in `constants.py`

On your next comment on readwise, please head to [https://readwise.io/books](https://readwise.io/books) to see your highlights. You should see them along with notes.

There are other crashes in the app (pre-existing), like trying to extract bookmarks before the audio file is converted to MP3.

**Comment by gostak-dd** [comment]
Also FYI I tried transcription with Whisper, there was hardly any difference. I don't use Audible that much anymore but for the few books I tested, the current transcription service works fine. If you are seeing issues then I welcome any solutions.

**Comment by gostak-dd** [comment]
Another thing, could you update the readme with Readwise info, such as where to get the token, and where to view the exported content?
